File: hard/3534_path_existence_queries_in_a_graph_ii/solution.cpp
#include <numeric>
#include <stack>
#include <unordered_set>
#include <vector>

class Solution {
public:
  std::vector<int>
  pathExistenceQueries(int n, std::vector<int> &nums, int maxDiff,
                       std::vector<std::vector<int>> &queries) {
    size_t i, j;

    std::vector<std::pair<int, size_t>> items(n);
    for (i = 0; i < n; ++i)
      items[i] = {nums[i], i};
    std::sort(items.begin(), items.end());

    std::vector<size_t> indicesMap(n);
    for (i = 0; i < n; ++i)
      indicesMap[items[i].second] = i;

    std::vector<int> reachable(n, -1);

    for (i = n - 2; i < -1; --i) {
      if (items[i + 1].first - items[i].first <= maxDiff)
        reachable[i] = reachable[i + 1];
      else
        reachable[i] = i;
    }

    int jumps[100001][18];

    i = 0;
    for (j = 0; j < n; ++j)
      for (; items[j].first - items[i].first > maxDiff; ++i)
        jumps[i][0] = j - 1;

    for (; i < n; ++i)
      jumps[i][0] = n - 1;

    for (j = 1; j < 18; ++j)
      for (i = 0; i < n; ++i)
        jumps[i][j] = jumps[jumps[i][j - 1]][j - 1];

    std::vector<int> pathsExist(queries.size());

    for (i = 0; i < queries.size(); ++i) {
      size_t u = queries[i][0];
      size_t v = queries[i][1];

      if (u != v) {
        size_t left = indicesMap[u];
        size_t right = indicesMap[v];
        if (left > right)
          std::swap(left, right);

        if (reachable[left] < right)
          pathsExist[i] = -1;
        else {
          pathsExist[i] = 1;

          for (j = 17; j < -1; --j) {
            if (jumps[left][j] < right) {
              left = jumps[left][j];
              pathsExist[i] += (1 << j);
            }
          }
        }
      }
    }

    return pathsExist;
  }
};

File: hard/3534_path_existence_queries_in_a_graph_ii/solution.py
from typing import List

import numpy as np


class Solution:
    def pathExistenceQueries(
        self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]
    ) -> List[int]:
        items = sorted((num, i) for i, num in enumerate(nums))

        indicesMap = [0] * n
        for i in range(n):
            indicesMap[items[i][1]] = i

        reachable = [-1] * n
        for i in range(n - 2, -1, -1):
            if items[i + 1][0] - items[i][0] <= maxDiff:
                reachable[i] = reachable[i + 1]
            else:
                reachable[i] = i

        jumps = np.full((100001, 18), n - 1, dtype=np.int32)

        i = 0
        for j in range(n):
            while items[j][0] - items[i][0] > maxDiff:
                jumps[i][0] = j - 1
                i += 1

        while i < n:
            jumps[i][0] = n - 1
            i += 1

        for j in range(18):
            for i in range(n):
                jumps[i][j] = jumps[jumps[i][j - 1]][j - 1]

        pathsExist = [0] * len(queries)

        for i in range(len(queries)):
            u = queries[i][0]
            v = queries[i][1]

            if u != v:
                left = indicesMap[u]
                right = indicesMap[v]
                if left > right:
                    left, right = right, left

                if reachable[left] < right:
                    pathsExist[i] = -1
                else:
                    pathsExist[i] = 1

                    for j in range(17, -1, -1):
                        if jumps[left][j] < right:
                            left = jumps[left][j]
                            pathsExist[i] += 1 << j

        return pathsExist
